docs/configure.rst
*********
Configure
*********

.. warning::
   The configuration values shown here are intended to demonstrate how to configure PEAT. **We STRONGLY recommend customizing the settings for your use case and environment.** The best example of this is limiting the PEAT device modules used to only those present in the environment, e.g. ``-d sel`` if using PEAT with SEL-manufactured devices.

Configuring PEAT
================
PEAT has a number of global settings that are configurable via several methods, including command line arguments, environment variables, and a :term:`YAML` file.

**If a value is configured via multiple methods, the order of precedence determines what value is actually used.**

.. _config-precedence:

Order of precedence for configuration
-------------------------------------
1. Command line arguments (example: ``-VV``)
2. Environment variables (example: ``export PEAT_DEBUG=2``)
3. Configuration file (example: ``-c peat-config.yaml``)
4. Default values

If using the Python API, e.g. ``config.DEBUG=1`` or values passed to ``initialize_peat(...)``, those are applied at the same level as command line arguments and override environment variables and the config file.


YAML configuration file
=======================
PEAT can be configured using a :term:`YAML` file. This file can be provided on the :term:`CLI` using the ``-c`` argument, for example ``peat scan -c peat-config.yaml``.

Values loaded from the file override the default values, but can be overridden by environment variables and command line arguments. In other words, options set in a config file have lower precedence than those set via command line args or environment variables.

Refer to :ref:`peat-config` for the available configuration options.

Config file usage examples
--------------------------
.. code-block:: bash

   peat scan -c ./examples/peat-config.yaml -d clx -i 192.0.2.0/24
   peat pull -c ./examples/peat-config.yaml -d ion sel -i 192.0.2.0/24
   peat push -c ./examples/peat-config.yaml -d selrelay -i 192.0.2.1 -- ./SET_1.TXT
   peat parse -c ./examples/peat-config.yaml -d selrelay ./SET_ALL.TXT


Walkthrough of the YAML config
==============================
The file starts with a ``metadata`` section. This is metadata about the config file itself, including the ``name`` of the config, a human-readable ``description``, the name of the original ``author``, timestamp of when it was ``created``, and timestamp of when it was ``updated``. ``name`` should be set, the others are optional but recommended.

.. code-block:: yaml
   :caption: Metadata example

   metadata:
     name: "simple-peat-config"
     description: "Simplified PEAT configuration with all comments removed"
     author: "cegoes"
     created: "May 22nd, 2024"
     updated: ""


Most of the top-level keys are standard config options, and most can be specified via command line args or environment variables, with a few exceptions (notable example are a lot of the elastic options aren't CLI-configurable to reduce complexity and size of ``--help``). Examples include ``resolve_ip``, ``no_print_results``, ``elastic_server``, ``debug``, and others.

.. code-block:: yaml
   :caption: Examples of standard config options

   verbose: false
   quiet: false
   no_print_results: false
   no_color: false
   no_logo: false
   assume_online: false
   max_threads: 260
   default_timeout: 5.0


The ``pillage`` section is the configuration for PEAT Pillage, refer to :ref:`pillage-config`.

The ``device_options`` section are settings for modules or protocols that are applied to everything in this PEAT run. In other words, it's global/universal, it's not on a per-host basis. Generally speaking, module/vendor specified methods have a dedicated key, e.g. ``sel`` for SEL devices (:class:`~peat.modules.sel.sel_relay.SELRelay` and :class:`~peat.modules.sel.sel_rtac.SELRTAC` modules), ``sage`` for the :class:`~peat.modules.schneider.sage.sage.Sage` module, etc. Then, there are protocol-specific options, e.g. ``telnet``, ``ssh``, etc. The protocol-specific options are usually port, timeout, login credentials, and any other protocol-specific options (such as SSH key paths).

.. code-block:: yaml
   :caption: Simple example forcing ``telnet`` to be used for pulls from any SEL devices.

   device_options:
   sel:
      pull_methods:
         - telnet


The ``hosts`` section is a list of hosts that will be scanned/pulled/interrogated by PEAT. Think of it as a inventory of devices. PEAT will use the information about these those to tune it's scanning parameters. Additionally, this is where per-host configurations are set, notably login credentials, as well as any other settings that need to be set for a particular host.

.. code-block:: yaml
   :caption: Example of the hosts section

   hosts:
   - label: "SEL-351S"
     comment: "SEL-351S Protection System in building XXX"
     identifiers:
       ip: 192.0.0.220
       mac: 00:30:A7:11:12:13
       serial_port: /dev/ttyUSB0
     peat_module: "SELRelay"
     options:
       ftp:
         user: "FTPUSER"
         pass: "TAIL"
   - label: "SEL-351"
     comment: "SEL-351 Protection System in building XXX"
     identifiers:
       ip: 192.0.0.221
       mac: 00:30:A7:11:12:14
     peat_module: "SELRelay"
     options:
       ftp:
         user: "FTP"
         pass: "TAIL"
       sel:
         never_download_dirs:
           - EVENTS


It's important to keep in mind that config options are not consistent. PEAT modules have evolved over time and they don't always follow the same way of doing things as other modules, especially those modules that existed before the YAML config was a thing (YAML config was introduced in late 2021). If you're uncertain of a option's behavior, refer to the reference config, and the module's class definition if needed.


Environment variables
=====================
Configuration options can be set via system environment variables that are prefixed with ``PEAT_``. For example, to change the debugging level to ``1``, set the environment variable ``PEAT_DEBUG`` to ``1``, such as with ``export PEAT_DEBUG=1`` on Linux. Environment variables will override default settings and settings loaded from a configuration file, and are overridden by command line arguments. For example, if the environment variable ``PEAT_DEBUG`` is set to ``1``, and peat is run with ``peat scan -VV``, then the value of :attr:`DEBUG <peat.settings.Configuration.DEBUG>` for that run of PEAT will be ``2``.

.. code-block:: bash
   :caption: Linux environment variable configuration example

   # Set variables via export
   export PEAT_DEBUG=1
   export PEAT_VERBOSE=true
   peat parse examples/
   peat scan -i localhost

   # Modify only for this command execution
   PEAT_DEBUG=2 peat parse examples/

.. code-block:: batch
   :caption: Windows environment variable configuration example

   setx PEAT_DEBUG 1
   setx PEAT_VERBOSE true
   peat parse examples/
   peat scan -i localhost


Additional topics
=================

Disabling file output
---------------------
Setting directory-related configuration options (e.g. :attr:`SUMMARIES_DIR <peat.settings.Configuration.SUMMARIES_DIR>`) to an empty string will **disable** any output to that directory. For example, setting :attr:`LOG_DIR <peat.settings.Configuration.LOG_DIR>` to an empty string (e.g. ``LOG_DIR=""``) will disable writing of logging data to files, including PEAT's logging and any protocols that log to a file (such as Telnet).

This feature is works well for most of the general options, such as :attr:`SUMMARIES_DIR <peat.settings.Configuration.SUMMARIES_DIR>` or :attr:`ELASTIC_DIR <peat.settings.Configuration.ELASTIC_DIR>`. Be warned, however, that mileage may vary for heavily used output dirs, such as :attr:`DEVICE_DIR <peat.settings.Configuration.DEVICE_DIR>` and :attr:`OUT_DIR <peat.settings.Configuration.OUT_DIR>`. They should work fine, but there have been regressions in the past where files have been written when :attr:`OUT_DIR <peat.settings.Configuration.OUT_DIR>` was disabled (in one case). If disabling file output is critical to your use case, we recommend testing locally before executing in the field.

Auto-generated configs
----------------------
Every run of PEAT generates a YAML file with the configuration values from the run in the path set by :attr:`META_DIR <peat.settings.Configuration.META_DIR>`, which by default is ``./peat_results/*/peat_metadata/``. This contains values for all configurations, regardless of their source, and is the single source of truth for how PEAT was configured at the end of a run. Note that if you used a config file for the run the auto-generated config may not match the config file you specified exactly.

These auto-generated configuration files can also be safely re-used in a future run without modification. For example, ``peat scan -c ./peat_results/*/peat_metadata/peat_configuration.yaml -i 192.0.2.0/24``. This can help ensure consistency between runs, simplifies the process of reproducibility (redoing the same run in the same manner at a later date), and saves typing.

JSON file
---------
PEAT will also accept configuration files in :term:`JSON` format. This is provided for flexibility and to provide backward compatibility. However, it's more limited than the YAML, and harder to write. The YAML format is preferred.


.. _peat-config:

YAML config reference
=====================
.. literalinclude:: ../examples/peat-config.yaml
   :name: PEAT YAML configuration reference
   :language: yaml


.. _pillage-config:

Pillage config reference
========================
- ``auto_copy`` [True or False] If ``True`` automatically copy a file to the results directory.  If ``False`` ask for permission to copy first.  This can be helpful in the case when there may be lot of false positives and the user wants to verify prior to copying. For example if you are searching for ``xml`` files.
- ``recursive`` [True of False] If ``True`` search in the source parent directory and all sub directories.  If ``False`` only search in the source parent directory and ignore all sub directories.
- Default/Brand specific search criteria:

  - ``locations`` [List of strings, each one should be a directory path] If empty search through all directories.  If directories exist only files in those directories will be considered. **Not yet implemented, Pillage currently searches through all directories regardless of what is listed here.**
  - ``filenames`` [List of strings] A list of filenames to search for.  When searching filenames take precedence over extensions and will be used first to determine the validity of a file.  No wildcards are accepted.  Must be the full filename with extension. Example ``['set_all.txt', '700g_001.rdb']``
  - ``extensions`` [List of strings] A list of extensions to search for.  No wildcards are accepted.  Example ``['txt', 'rdb', 'xml']``


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
TBD


peat/settings.py
from __future__ import annotations

import typing
from collections import defaultdict
from ipaddress import IPv4Interface, IPv4Network
from pathlib import Path

from .consts import RUN_ID, START_TIME, AllowedCommTypes, EntrypointType
from .settings_manager import SettingsManager

if typing.TYPE_CHECKING:
    try:
        import peat
    except Exception:
        pass


class Configuration(SettingsManager):
    """
    Global configuration settings used throughout PEAT.
    """

    METADATA: dict = {
        "name": "",
        "description": "",
        "author": "",
        "created": "",
        "updated": "",
    }
    """
    Metadata for the PEAT YAML configuration file.
    """

    ADDITIONAL_MODULES: list[str | Path] = []
    """
    File paths to external/third-party PEAT device modules to import.

    These can be the path to a ``.py`` file with a :class:`~peat.device.DeviceModule`
    subclass or a path to a folder with an ``__init__.py`` file and any number
    of ``.py`` files containing :class:`~peat.device.DeviceModule` subclasses
    to import.
    """

    DEBUG: int = 0
    """
    DEBUG level (higher = more output, 0 = disabled).
    """

    VERBOSE: bool = False
    """
    Include DEBUG messages in the :term:`CLI` terminal output.
    """

    QUIET: bool = False
    """
    Don't write log messages to the :term:`CLI` terminal.
    """

    # TODO: delete in a future release (probably in 2025)
    NO_PRINT_RESULTS: bool = False
    """
    DEPRECATED. Setting this no longer has any effect,
    as it's now the default behavior.
    """

    PRINT_RESULTS: bool = False
    """
    Print JSON-formatted results from the operation to
    the terminal (stdout). Note that log messages will still
    be printed unless '--quiet' is specified.
    """

    NO_COLOR: bool = False
    """
    Don't color log messages in the :term:`CLI` terminal.
    """

    NO_LOGO: bool = False
    """
    Don't print the PEAT logo at startup in the :term:`CLI` terminal.
    """

    DRY_RUN: bool = False
    """
    "Dry run" when running on the CLI, do everything except running commands or
    connecting to servers. Actions won't be executed, but logs and state will
    still be written to files and saved to Elasticsearch (if enabled).
    This includes loading and verifying configuration and importing modules

    Useful for verifying a YAML config file or other settings are correct,
    or that a third-party PEAT module is imported correctly.
    """

    ASSUME_ONLINE: bool = False
    """
    Skip the host online check before scan/pull/push.
    """

    MAX_THREADS: int = 260
    """
    Maximum number of threads for any concurrent operations (scanning, etc.).
    """

    DEFAULT_TIMEOUT: float = 5.0
    """
    Default timeout for sockets and potentially other things.
    """

    RESOLVE_IP: bool = True
    """
    If PEAT should attempt to resolve device :term:`IP`
    address from a hostname or :term:`MAC` address.
    """

    RESOLVE_MAC: bool = True
    """
    If PEAT should attempt to resolve device :term:`MAC`
    address from it's :term:`IP` address.
    """

    RESOLVE_HOSTNAME: bool = True
    """
    If PEAT should attempt to resolve device hostname
    from it's :term:`IP` address.
    """

    FORCE_ONLINE_METHOD_PING: bool = False
    """
    Force :term:`ARP` and :term:`ICMP` requests to be used to check if a host
    is online, even if the system running PEAT isn't able to use them.
    """

    FORCE_ONLINE_METHOD_TCP: bool = False
    """
    Force :term:`TCP` SYNs to be used to check if a host is online, even if
    the system running PEAT is able to use :term:`ARP` or :term:`ICMP` to
    perform the checks.
    """

    ICMP_FALLBACK_TCP_SYN: bool = True
    """
    In the case of a :term:`ICMP` failure, fallback to attempting a
    :term:`TCP` SYN RST to check if the host is online. If false,
    then :term:`ICMP` failures will result in the host being marked
    as down, even if they're blocked by a firewall or gateway.
    """

    SYN_PORT: int = 80
    """
    Default port used for basic :term:`TCP` SYN online checks.
    The default :term:`HTTP` (web) port 80 is generally safe to check.
    """

    PUSH_SKIP_SCAN: bool = False
    """
    Skip scanning and verification of hosts being pushed to and assume
    all hosts are online and valid devices.

    .. note::
       This requires a single device type to be specified
       (the ``-d`` argument on the :term:`CLI`)
    """

    SCAN_SWEEP: bool = False
    """
    Simple host up/down check (equivalent to ``nmap -Pn <hosts>``).

    If serial ports are targeted, this will enumerate the active
    serial ports on the host.
    """

    INTENSIVE_SCAN: bool = False
    """
    Force identification checks of all ports during scanning.
    """

    HASH_ALGORITHMS: list[str] = ["md5", "sha1", "sha256", "sha512"]
    """
    Hash algorithms to use wherever hashes are calculated.
    Available algorithms are any provided by :mod:`hashlib`.
    """

    OUT_DIR: Path = Path.cwd() / "peat_results"
    """
    Default directory for all file output.

    :meta hide-value:
    """

    RUN_DIR: Path = OUT_DIR / f"default-run-dir_{START_TIME}_{RUN_ID}"
    """
    Output directory for all files associated with a single run of PEAT.

    Name format: ``<command>_<config-name>_<timestamp>_<run-id>``

    - ``<command>`` : PEAT command, e.g. "scan", "pull", "parse", etc.
    - ``<config-name>`` : name of YAML config file, set in ``metadata: name: "name"``.
        If no config name is specified, then the string ``default-config`` is used.
    - ``<timestamp>`` : start time of the PEAT run, e.g. ``2022-06-15_13-08-59``.
        This value is retrieved from ``consts.START_TIME``.
    - ``<run-id>`` : Run ID, aka ``agent.id``, e.g. ``165532013980``.
        This value is retrieved from ``consts.RUN_ID``.

    Examples:

    - ``pull_sceptre-test-config_2022-06-17_165532013980``
    - ``scan_default-config_2022-09-27_165532013980``

    :meta hide-value:
    """

    DEVICE_DIR: Path = RUN_DIR / "devices"
    """
    Output directory for files generated by PEAT device modules,
    e.g. pulled config files, firmware images, etc.
    This is the most relevant directory for the majority of PEAT users.

    .. note::
       Device file output can be disabled entirely by setting this to
       a empty string (``""``) or :obj:`None`. This can be useful for
       low-footprint use cases.

    :meta hide-value:
    """

    ELASTIC_DIR: Path = RUN_DIR / "elastic_data"
    """
    Directory where raw documents pushed to Elasticsearch are saved in JSON format.
    These can be used to rebuild the Elasticsearch indices if needed.

    .. note::
       Elastic doc saving can be disabled entirely by setting this to
       a empty string (``""``) or :obj:`None`.

    :meta hide-value:
    """

    LOG_DIR: Path = RUN_DIR / "logs"
    """
    Directory for PEAT's log files.
    Setting this to an empty string will disable output of these files.

    :term:`JSON` formatted log files are also saved here.
    These can be used to rebuild the ``vedar-logs`` Elasticsearch index.

    Files that may be created in this directory include:

    - enip/* (Rockwell-specific)
    - peat.log
    - json-log.jsonl
    - elasticsearch.log
    - telnet.log

    :meta hide-value:
    """

    META_DIR: Path = RUN_DIR / "peat_metadata"
    """
    Directory for PEAT's run metadata, notably dumps of PEAT's
    configuration and internal state.

    Setting this to an empty string will disable output of these files.

    .. note::
       Log files are stored separately, in ``logs/``

    Files that may be created in this directory include:

    - peat_configuration.yaml
    - peat_configuration.json
    - peat_state.json
    - Copy of the config file used for the run,
        if applicable (e.g. ``-c my-config.yaml``)

    :meta hide-value:
    """

    SUMMARIES_DIR: Path = RUN_DIR / "summaries"
    """
    Directory for summary results for PEAT commands.

    Files that may be created in this directory include:

    - scan-summary.json
    - pull-summary.json
    - parse-summary.json

    :meta hide-value:
    """

    TEMP_DIR: Path = RUN_DIR / "temp"
    """
    Working directory for file artifacts.
    May be deleted when PEAT exits.

    :meta hide-value:
    """

    ZEEK_LOGDIR: Path = RUN_DIR / "zeek_logs"
    """
    Directory where Zeek logs will be saved. Also contains Zeek artifacts.
    """

    ELASTIC_SERVER: str = None
    """
    URL of the Elasticsearch server.
    """

    ELASTIC_SAVE_LOGS: bool = True
    """
    If PEAT logs should be sent to Elasticsearch.
    """

    ELASTIC_SAVE_CONFIG: bool = True
    """
    If PEAT Configuration dumps should be sent to Elasticsearch on exit.
    """

    ELASTIC_SAVE_STATE: bool = True
    """
    If PEAT State dumps should be sent to Elasticsearch on exit.
    """

    ELASTIC_SAVE_BLOBS: bool = False
    """
    If large binary blobs should be stored in Elasticsearch
    (e.g. firmware images).
    """

    ELASTIC_TIMEOUT: float = 10.0
    """
    Timeout to connect to the Elasticsearch server.
    """

    ELASTIC_DISABLE_DATED_INDICES: bool = False
    """
    Append the current date to Elasticsearch index names.

    Example: when false, a push to ``peat-configs`` on Feb 25 2021
    will actually push to an index named ``peat-configs-2021.02.25``.
    When this option disabled (set to True), the push will instead
    go to ``peat-configs`` (as above, without the date).
    """

    ELASTIC_ADDITIONAL_TAGS: list[str] = []
    """
    Additional tag strings to add to the ``tags`` field for
    each document pushed to Elasticsearch.
    """

    ELASTIC_LOG_INDEX: str = "vedar-logs"
    """
    Base name of Elasticsearch index to use PEAT logging events.
    """

    ELASTIC_SCAN_INDEX: str = "peat-scan-summaries"
    """
    Base name of Elasticsearch index to use for scan result summaries,
    e.g. what normally gets written in ``peat_results/<run-dir>/summaries/``.
    """

    ELASTIC_PULL_INDEX: str = "peat-pull-summaries"
    """
    Base name of Elasticsearch index to use for pull result summaries,
    e.g. what normally gets written in ``peat_results/<run-dir>/summaries/``.
    """

    ELASTIC_PARSE_INDEX: str = "peat-parse-summaries"
    """
    Base name of Elasticsearch index to use for parse result summaries,
    e.g. what normally gets written in ``peat_results/<run-dir>/summaries/``.
    """

    ELASTIC_CONFIG_INDEX: str = "peat-configs"
    """
    Base name of Elasticsearch index to use for PEAT configuration dumps
    from runs of PEAT, e.g. what normally gets written in
    ``peat_results/<run-dir>/metadata/``.
    """

    ELASTIC_STATE_INDEX: str = "peat-state"
    """
    Base name of Elasticsearch index to use for PEAT state dumps
    from runs of PEAT, e.g. what normally gets written in
    ``peat_results/<run-dir>/metadata/``.
    """

    ELASTIC_HOSTS_INDEX: str = "ot-device-hosts-timeseries"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.DeviceData`
    data in 'timeseries' form (new entries are created for every collection
    performed or data parsed).

    Information collected by PEAT from field devices or parsed files.
    A new Elasticsearch document is created for every pull of datafrom a device
    (the data is 'timeseries', with differences visible between pulls over time).
    """

    ELASTIC_FILES_INDEX: str = "ot-device-files"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.File` data.

    Information about files present on the device, or that were present on the
    device at one point in time.
    """

    ELASTIC_REGISTERS_INDEX: str = "ot-device-registers"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.Register` data.

    Information about individual communication 'registers' (e.g. Modbus
    registers/coils, DNP3 data points, BACNet objects, etc.) that are
    configured on devices, as extracted from device configuration
    information.
    """

    ELASTIC_TAGS_INDEX: str = "ot-device-tags"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.Tag` data.

    Information about tag variables that are configured on devices,
    as extracted from device configuration information.
    """

    ELASTIC_IO_INDEX: str = "ot-device-io"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.IO` data.

    Information about I/O (Input/Output) available and/or configured
    on a device, as extracted from device configuration information.
    """

    ELASTIC_EVENTS_INDEX: str = "ot-device-events"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.Event`
    data from devices.

    Logging and other event history as extracted from devices.
    Examples include access logs, system logs, or protection history
    """

    ELASTIC_MEMORY_INDEX: str = "ot-device-memory"
    """
    Base name of Elasticsearch index to use for :class:`~peat.data.models.Memory`
    data from devices.

    Memory reads from devices, including address in memory, the value read,
    and information about where it came from and when the read occurred.
    """

    ELASTIC_UEFI_FILES_INDEX: str = "uefi-files"
    """
    Base name of elasticsearch index to use for :class:`peat.data.models.UEFIFiles`

    Information is parsed from a file
    """

    ELASTIC_UEFI_HASHES_INDEX: str = "uefi-hashes"
    """
    Base name of elasticsearch index to use for :class:`peat.data.models.UEFIHash`
    """
    HEAT_ELASTIC_SERVER: str = None
    """
    Elasticsearch server to pull :term:`HEAT` data from.
    """

    HEAT_INDEX_NAMES: str = "packets-*"
    """
    Names and/or patterns of Elasticsearch indices with Packetbeat
    data for :term:`HEAT`.
    """

    HEAT_DATE_RANGE: str = None
    """
    Date range to filter :term:`HEAT` extraction to.
    """

    HEAT_EXCLUDE_IPS: list[str] = []
    """
    IP addresses to exclude from packetbeat search
    (source and/or destination IP).
    """

    HEAT_ONLY_IPS: list[str] = []
    """
    IP address to limit packetbeat search to
    (source or destination IP).
    """

    HEAT_FILE_ONLY: bool = False
    """
    Extract the file(s) but don't parse them using PEAT.
    """

    HEAT_ARTIFACTS_DIR: Path = RUN_DIR / "heat_artifacts"
    """
    Directory where :term:`HEAT` artifacts should be saved.

    :meta hide-value:
    """

    PILLAGE: dict = {}
    """
    Configuration for ``peat pillage`` (:ref:`pillage`).
    """

    DEVICE_OPTIONS: dict = {}
    """
    Protocol and module configuration options that apply to all hosts.
    """

    HOSTS: list[dict] = []
    """
    Specify or override options for specific hosts.
    """

    PCAPS: Path | None = None
    """
    Specify folder that contains PCAPS for processing by HEAT.

    :meta hide-value:
    """

    HEAT_PROTOCOLS: list = []
    """
    Specify protocols for HEAT to use.
    """

    NO_RUN_ZEEK: bool = False
    """
    Don't have PEAT run Zeek on a PCAP, instead run it
    on a PCAP file, then process the output of the PCAP.

    NOTE: zeek_dir must be specified if no_run_zeek is true.
    """

    ZEEK_DIR: Path | None = None
    """
    Directory with existing Zeek output to use as input.

    This is an alternative to PEAT running Zeek on the PCAPs itself.

    NOTE: This argument is required if no_run_zeek is true.

    :meta hide-value:
    """


class State(SettingsManager):
    """
    Global persistent state singleton, e.g. caching results of expensive operations.
    """

    arp_table: str = ""
    """
    The host's ARP table. This is updated if a MAC address lookup is made and the ARP
    table is queried by PEAT. The ARP table is not always queried during lookups,
    so this variable may not always be populated. It also won't be populated if
    ``RESOLVE_MAC`` is False.
    """

    comm_type: AllowedCommTypes | None = None
    """
    Communication method being used for an active command (scan, pull, or push).
    This is a bit of a hack to allow leaking a bit of CLI-specific info
    to device modules that need it.

    Allowed values: ``unicast_ip``, ``broadcast_ip``, ``serial``

    :meta hide-value:
    """

    elastic: peat.Elastic | None = None
    """
    Elastic instance used to interact with a Elasticsearch database,
    such as the one used by Vedizar.

    :meta hide-value:
    """

    local_interface_names: list[str] = []
    """
    Names of local network interfaces on the current system, e.g. ``eth0``.

    :meta hide-value:
    """

    local_interface_details: dict[str, dict[str, str]] = {}
    """
    Detailed local interface IPv4 and link configuration information.

    :meta hide-value:
    """

    local_interface_ips: list[str] = []
    """
    :term:`IP` addresses of local interfaces.

    :meta hide-value:
    """

    local_interface_macs: list[str] = []
    """
    :term:`MAC` addresses of local interfaces.

    :meta hide-value:
    """

    local_interface_objects: list[IPv4Interface] = []
    """
    :class:`~ipaddress.IPv4Interface` objects of the local interfaces.

    :meta hide-value:
    """

    local_interface_networks: dict[str, list[IPv4Network]] = defaultdict(list)
    """
    Mapping of local interface names to their associated subnets.

    :meta hide-value:
    """

    local_networks: list[IPv4Network] = []
    """
    Networks the local system is connected to.

    :meta hide-value:
    """

    default_interface_name: str = ""
    """
    The name of the default network interface of the local system.

    :meta hide-value:
    """

    default_interface_ip: str = ""
    """
    The :term:`IP` address of the default network interface of the local system.

    :meta hide-value:
    """

    default_interface_mac: str = ""
    """
    The :term:`MAC` address of the default network interface of the local system.

    :meta hide-value:
    """

    default_gateway: str = ""
    """
    The :term:`IP` address of the default gateway of the local system.

    :meta hide-value:
    """

    raw_socket_capable: bool = False
    """
    If RAW sockets are available (e.g. for :term:`ARP` or :term:`ICMP` requests).

    :meta hide-value:
    """

    entrypoint: EntrypointType = "Package"
    """
    How PEAT is being used (PEAT's "entrypoint").

    Allowed values:

    - CLI: PEAT's CLI internals.
    - Server: if being run as a server, e.g. web UI.
    - Package: any other use of PEAT's API.

    :meta hide-value:
    """

    peat_initialized: bool = False
    """
    If PEAT has been initialized, in other words
    :func:`peat.init.initialize_peat` has been called at least once.

    :meta hide-value:
    """

    written_files: set[str] = set()
    """
    Paths of files written by PEAT during the current run.
    Only includes files that were written using PEAT's APIs.

    .. note::
       These are NOT ordered by when they were written! (it's a :class:`set`)

    :meta hide-value:
    """

    superuser_privs: bool | None = None
    """
    Is PEAT running with permissions of ``root`` (POSIX)
    or ``Administrator`` (Windows).

    :meta hide-value:
    """

    error: bool = False
    """
    If any critical errors occurred during the run of PEAT.

    :meta hide-value:
    """


config = Configuration(label="configuration", env_prefix="PEAT_", init_env=True)
state = State(label="state", env_prefix="PEAT_STATE_", init_env=True)


__all__ = ["Configuration", "State", "config", "state"]
